### PR TITLE
Allow users to add metrics to Prometheus via StatsD

### LINKED ISSUE #1897: Support exporting custom Prometheus metrics from predictor implementations

### PR DESCRIPTION
**Pull request by miguelvr**
closes #1897 

---

checklist:

- [x] run `make test` and `make lint`
- [x] test manually (i.e. build/push all images, restart operator, and re-deploy APIs)

### CODE REVIEW

File: pkg/cortex/serve/cortex_internal/lib/metrics.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+from datadog import DogStatsd
+
+from cortex_internal.lib.exceptions import UserException
+
+
+def validate_metric(fn):
+    def _metric_validation(self, metric: str, value: float, tags: Dict[str, str] = None):
+        validate_metric_name(metric)
+        return fn(self, metric=metric, value=value, tags=transform_tags(tags))
```
Comment (deliahu) [inline pkg/cortex/serve/cortex_internal/lib/metrics.py:25]: Consider moving `transform_tags` to within the function

File: pkg/cortex/serve/start/batch.py [hunk 2]
```diff
@@ -330,7 +332,13 @@ def start():
 
     try:
         logger.info("loading the predictor from {}".format(api.predictor.path))
-        predictor_impl = api.predictor.initialize_impl(project_dir, client, job_spec)
+        metrics_client = MetricsClient(api.statsd)
```
Comment (deliahu) [inline pkg/cortex/serve/start/batch.py:335]: Do we support this for tasks?

**Review by deliahu — APPROVED** [review]

File: pkg/cortex/serve/cortex_internal/lib/api/api.py [hunk 3]
```diff
@@ -50,7 +51,11 @@ def __init__(
 
         host_ip = os.environ["HOST_IP"]
         datadog.initialize(statsd_host=host_ip, statsd_port=9125)
-        self.statsd = datadog.statsd
+        self.__statsd = datadog.statsd
```
Comment (vishalbollu) [inline pkg/cortex/serve/cortex_internal/lib/api/api.py:54]: nit: IIRC, single underscore is for indicating private variables. Is there a convention for double underscore?

File: pkg/cortex/serve/cortex_internal/lib/api/task.py [hunk 4]
```diff
@@ -92,38 +129,10 @@ def _read_impl(self, module_name: str, impl_path: str, target_class_name: str):
...
-                    )
-            else:
-                raise UserException(
-                    f"class {target_class_name}",
-                    f'"__call__" method not defined',
-                )
+            validate_class_impl(impl, TASK_CLASS_VALIDATION)
         else:
             callable_fn = impl
             argspec = inspect.getfullargspec(callable_fn)
```
Comment (vishalbollu) [inline pkg/cortex/serve/cortex_internal/lib/api/task.py:138]: If the user provides a callable (i.e. function), then they wouldn't be able to use the metrics client?

File: pkg/cortex/serve/cortex_internal/lib/api/task.py [hunk 5]
```diff
@@ -12,16 +12,36 @@
...
+from cortex_internal.lib.metrics import MetricsClient
 
 logger = configure_logger("cortex", os.environ["CORTEX_LOG_CONFIG_FILE"])
 
+TASK_CLASS_VALIDATION = {
+    "required": [
+        {
+            "name": "__init__",
+            "required_args": ["self"],
+            "optional_args": ["metrics_client"],
```
Comment (vishalbollu) [inline pkg/cortex/serve/cortex_internal/lib/api/task.py:34]: We can pass in the `metrics_client` in the `__call__` function as opposed to the `init` function. In the current design both `__init__` and `__call__` will be invoked once.

**Review by vishalbollu — COMMENTED** [review]
Let's discuss the Task design changes

File: pkg/cortex/serve/cortex_internal/lib/api/api.py (hunk 3 above)
Comment (miguelvr) [reply pkg/cortex/serve/cortex_internal/lib/api/api.py:54]: double underscore throws an error if you try to use it

File: pkg/cortex/serve/cortex_internal/lib/api/task.py (hunk 4 above)
Comment (miguelvr) [reply pkg/cortex/serve/cortex_internal/lib/api/task.py:138]: I opted for that approach for consistency with the rest of the APIs.

The class version becomes the "advanced use case" while the function becomes the straightforward one

**Review by vishalbollu — APPROVED** [review]

**Pull request merged**